toqito/state_opt/npa_hierarchy.py

```diff
@@ -157,129 +263,172 @@ def npa_constraints(
     dim = len(words)
 
-    r_var = cvxpy.Variable((referee_dim * dim, referee_dim * dim), hermitian=True, name="R")
-    # Normalization.
-    norm = sum(r_var[i * dim, i * dim] for i in range(referee_dim))
-    constraints = [norm == 1, r_var >> 0]
+    if dim == 0:
+        # Should not happen if IDENTITY_SYMBOL is always included
+        raise ValueError("Generated word list is empty. Check _gen_words logic.")
+
+    # Moment matrix (Gamma matrix in NPA paper)
+    # r_var[i,j] block corresponds to E[S_i^dagger S_j]
+    moment_matrix_R = cvxpy.Variable((referee_dim * dim, referee_dim * dim), hermitian=True, name="R")
+
+    # Referee's effective state rho_R = E[I]
+    # This is the (0,0) block of moment_matrix_R since words[0] is Identity
+    rho_R_referee = moment_matrix_R[0:referee_dim, 0:referee_dim]
+
+    # Ensure rho_R_referee is a valid quantum state
+    constraints = [
+        cvxpy.trace(rho_R_referee) == 1,
+        rho_R_referee >> 0,
+        moment_matrix_R >> 0,
+    ]
+
+    # Store relations for (S_i^dagger S_j) -> block_index in moment_matrix_R
+    # This helps enforce Γ(S_i^dagger S_j) = Γ(S_k^dagger S_l) if products are algebraically equal
+    seen_reduced_products = {}
 
-    seen = {}
     for i in range(dim):
-        for j in range(i, dim):
-            w_i, w_j = words[i], words[j]
-            w_i = tuple(reversed(w_i))
-            word = _reduce(w_i + w_j)
-
-            sub_mat = r_var[i::dim, j::dim]
-            # if i = 0 we would consider (ε, ε) as an empty word.
-            if i != 0 and _is_zero(word):
-                constraints.append(sub_mat == 0)
-
-            elif _is_meas(word):
-                s_a, s_b = word
+        for j in range(i, dim):  # Iterate over upper triangle including diagonal
+            word_i_conj = tuple(reversed(words[i]))  # S_i^dagger
+
+            # The product S_i^dagger S_j
+            # For _reduce, ensure no IDENTITY_SYMBOL unless it's the only element.
+            # If word_i_conj is (ID,), S_i_dagger_S_j is S_j. If word_j is (ID,), it's S_i_dagger.
+            # If both are (ID,), product is (ID,).
+
+            product_unreduced = []
+            if word_i_conj != (IDENTITY_SYMBOL,):
+                product_unreduced.extend(list(word_i_conj))
+            if words[j] != (IDENTITY_SYMBOL,):
+                product_unreduced.extend(list(words[j]))
+
+            # This happens if both words[i] and words[j] were IDENTITY_SYMBOL
+            if not product_unreduced:
+                product_S_i_adj_S_j = (IDENTITY_SYMBOL,)
+            else:
+                product_S_i_adj_S_j = _reduce(tuple(product_unreduced))
+
+            # Moment matrix (Gamma matrix in NPA paper :cite:`Navascues_2008_AConvergent` - arXiv:0803.4290)
+            # This hierarchy can be generalized, e.g., to incorporate referee systems
+            # as seen in extended nonlocal games (see, e.g., F. Speelman's thesis, arXiv:1605.07720).
```

Comment:
This arXiv link is not correct (as far as I know, this work is not on the arXiv). We want to add the following entry into the `refs.bib` file:

```
@book{Speelman_2016_Position,
  title={Position-based quantum cryptography and catalytic computation},
  author={Speelman, Florian},
  year={2016},
  publisher={University of Amsterdam}
}
```

And then, subsequently:
```suggestion
            # as seen in extended nonlocal games (see, e.g., F. Speelman's thesis, :cite:`Speelman_2016_Position`).
```